iOS password support

brotskydotcom opened the pull request:
This PR adds support for setting/getting generic and internet passwords on iOS (and macOS) without use of the macOS file-based (legacy) keychain APIs.  It does most of the work to fix #70, although it doesn't go all the way in terms of adding complete support for finding/returning all attributes of all items.

Because there's no way to actually test iOS support without using an iOS app, this PR also adds a Swift app with tests to the workspace tree, as well as a shim-library crate with a CoreFoundation interface that can be called directly from Swift.

It should be possible to enhance the GitHub CI action to do iOS as well as macOS testing, but this PR doesn't do that.

brotskydotcom commented:
@kornelski I believe this is ready for a first review.

kornelski commented on security-framework/src/item.rs:
```diff
@@ -175,7 +174,8 @@ impl ItemSearchOptions {
     /// Sets kSecAttrAccessGroup to kSecAttrAccessGroupToken
     #[inline(always)]
     pub fn access_group_token(&mut self) -> &mut Self {
-        self.access_group = unsafe { Some(CFString::wrap_under_get_rule(kSecAttrAccessGroupToken)) };
```
Please don't change formatting of unrelated things

brotskydotcom replied:
Sorry!  I use `rustfmt` and it made some changes which I mistakenly assumed were expected.  I will revert any changes it made in older code.

brotskydotcom replied:
Reverted in my latest commits.

kornelski commented on ios-shim/src/iospw.h:
```diff
@@ -0,0 +1,14 @@
+
+#include <CoreFoundation/CoreFoundation.h>
+
+extern OSStatus SecSetGenericPassword(CFStringRef service, CFStringRef account, CFDataRef password);
```
Can you add comments why this shim is needed?

brotskydotcom replied:
Yep, will do.

kornelski commented on ios-shim/src/lib.rs:
```diff
@@ -0,0 +1,60 @@
+use security_framework_sys::base::{errSecBadReq, errSecSuccess};
+
+#[no_mangle]
+pub extern "C" fn SecSetGenericPassword(
```
This uses Apple's `Sec*` naming scheme. All symbols in C are global, so there's a risk of a collision with Apple's functions. If this function is just for Rust, please add a custom name prefix, e.g. `rust_sec_framework_set_generic_password`

brotskydotcom replied:
This function is actually just for use from Swift and Objective C/C++ code, both of which use camel case naming.  I am happy not to use Apple's naming convention, but would prefer to keep using camel case.  How about `Rust` or `RustSec` or `RustShim` as a prefix, as in `RustShimSetGenericPassword`?  Do you have a preference?

kornelski replied:
RustShim is fine.

brotskydotcom replied:
OK, will use `RustShim`

kornelski commented on security-framework/src/passwords.rs:
```diff
@@ -0,0 +1,351 @@
+    if port.is_some() {
+        query.push((
+            unsafe { CFString::wrap_under_get_rule(kSecAttrPort) },
+            CFNumber::from(port.unwrap() as i32).as_CFType(),
```
Please use `if let` to avoid unwrap.

brotskydotcom replied:
Great catch, thanks.

brotskydotcom commented:
@kornelski I've addressed your review comments, added docs, and fixed the `clippy` issues that were breaking the build.  I think this is ready to go.  Let me know if you want me to enhance the CI script to run the iOS tests via XCode.

kornelski commented on security-framework/src/passwords.rs:
```diff
@@ -0,0 +1,361 @@
+    ));
+    let params = CFDictionary::from_CFType_pairs(&query);
+    let mut ret: CFTypeRef = std::ptr::null();
+    cvt(unsafe { SecItemCopyMatching(params.as_concrete_TypeRef(), &mut ret) })?;
```
`ret` needs to be retained using `wrap_under_create_rule`, but `get_password_internal` doesn't do that.

brotskydotcom replied:
Yep, same as above.

brotskydotcom commented on security-framework/src/passwords.rs:
```diff
@@ -0,0 +1,361 @@
+    ));
+    let params = CFDictionary::from_CFType_pairs(&query);
+    let mut ret: CFTypeRef = std::ptr::null();
+    cvt(unsafe { SecItemCopyMatching(params.as_concrete_TypeRef(), &mut ret) })?;
```
Great catch, thanks!  Resolved in latest commit, which adds a number of safety checks to `get_password_and_release` (renamed from `get_password_internal`).

brotskydotcom commented:
@kornelski While working on this PR, there were some clippy warnings in older code.  I'll open a separate issue with those.

kornelski commented:
Thank you
